### PR TITLE
schannel: add support for CURLOPT_CAINFO.

### PR DESCRIPTION
**Pull request by mcnulty**
__Modifications:__
- make verify_certificate functionality in curl_schannel.c
  available on all versions of Windows instead of just
  Windows CE. verify_certificate will only be invoked on
  Windows CE or when the user specifies CURLOPT_CAINFO and
  CURLOPT_SSL_VERIFYPEER
- In verify_certificate, create a custom certificate chain
  engine that exclusively trusts the certificate store backed
  by the CURLOPT_CAINFO file
- HAVE_EXCLUSIVE_TRUST_MODE macro defined in config-win32.h for
  Windows 7+ to control enabling this feature that relies on
  Windows 7+ functionality
- doc updates of --cacert/CAINFO support for schannel
- Use CERT_NAME_SEARCH_ALL_NAMES_FLAG when invoking CertGetNameString
  when available. This implements a TODO in schannel.c to improve
  handling of multiple SANs in a certificate. In particular, all SANs
  will now be searched instead of just the first name.
- Add new test cases 314 and 315. These test cases check that
  the first and last SAN, respectively, matches the connection
  hostname. New test certificates have been added for these
  cases. For 314, the certificate prefix is Server-localhost-firstSAN
  and for 315, the certificate prefix is Server-localhost-secondSAN
- Remove TODO 15.2 (Add support for custom server certificate
  validation), this commit addresses it.

__Testing__

I didn't have any luck running the test suite for this particular platform/lib configuration so I validated the following test cases by hand:
- test305
- test310
- test311
- test312
- test314
- test315
- test2043

All the tests I validated (except for test305) required specifying --ssl-no-revoke because the test CA (EdelCurlRoot-ca) does not have any revocation checking data in its cert and SChannel requires revocation checking by default.

I also ran the testsuite (for good measure) on Ubuntu 16.04.2/x86_64, using ./configure without args (built against OpenSSL) and didn't see any test failures.

### CODE REVIEW

**Comment by bagder** [comment]
The failed CI test is test 2033, which is yet another pipelining test we probably need to disable due to its flakiness.

File: lib/vtls/schannel.c [hunk 1]
```diff
@@ -83,9 +83,25 @@ static Curl_recv schannel_recv;
 static Curl_send schannel_send;
 
 #ifdef _WIN32_WCE
-static CURLcode verify_certificate(struct connectdata *conn, int sockindex);
+/* certificate validation on CE doesn't seem to work right; we'll
+ * do it following a more manual process. */
+static const int DEFAULT_MANUAL_CRED_VALIDATION = 1;
```
Comment (mback2k) [inline lib/vtls/schannel.c]: Please use a #define instead of a static const int variable that is not really needed in my opinion.

File: lib/vtls/schannel.c [hunk 2]
```diff
@@ -83,9 +83,25 @@ static Curl_recv schannel_recv;
...
+#else
+static const int DEFAULT_MANUAL_CRED_VALIDATION = 0;
```
Comment (mback2k) [inline lib/vtls/schannel.c]: Same as above, please use a compile-time #define instead.

File: lib/vtls/schannel.c [hunk 3]
```diff
@@ -83,9 +83,25 @@ static Curl_recv schannel_recv;
...
+#endif
+
+#ifdef HAVE_EXCLUSIVE_TRUST_MODE
+static const int CAFILE_SUPPORTED = 1;
```
Comment (mback2k) [inline lib/vtls/schannel.c]: Same as above, please use a compile-time #define instead.

File: lib/vtls/schannel.c [hunk 4]
```diff
@@ -83,9 +83,25 @@ static Curl_recv schannel_recv;
...
+#else
+static const int CAFILE_SUPPORTED = 0;
```
Comment (mback2k) [inline lib/vtls/schannel.c]: Same as above, please use a compile-time #define instead.

File: lib/vtls/schannel.c [hunk 5]
```diff
@@ -83,9 +83,25 @@ static Curl_recv schannel_recv;
...
 #endif
 
+static const int MAX_CAFILE_SIZE = 67108864; /* 64 MB */
+static const char * const BEGIN_CERT = "-----BEGIN CERTIFICATE-----\n";
+static const char * const END_CERT = "\n-----END CERTIFICATE-----";
```
Comment (mback2k) [inline lib/vtls/schannel.c]: Same as above, please use a compile-time #defines instead.

File: lib/vtls/schannel.c [hunk 6]
```diff
@@ -383,7 +402,7 @@ schannel_connect_step1(struct connectdata *conn, int sockindex)
   }
   memset(connssl->ctxt, 0, sizeof(struct curl_schannel_ctxt));
 
-  host_name = Curl_convert_UTF8_to_tchar(hostname);
+  host_name = Curl_convert_UTF8_to_tchar(conn->host.name);
```
Comment (mback2k) [inline lib/vtls/schannel.c]: I think this change breaks HTTPS-proxy support. Why was this changed from hostname which is used in all other places regarding certificate verification in this function?

File: lib/vtls/schannel.c [hunk 7]
```diff
@@ -558,7 +577,7 @@ schannel_connect_step2(struct connectdata *conn, int sockindex)
     memcpy(inbuf[0].pvBuffer, connssl->encdata_buffer,
            connssl->encdata_offset);
 
-    host_name = Curl_convert_UTF8_to_tchar(hostname);
+    host_name = Curl_convert_UTF8_to_tchar(conn->host.name);
```
Comment (mback2k) [inline lib/vtls/schannel.c]: Same as above, this might break HTTPS-proxy support. Please elaborate on the reason for this change.

File: lib/vtls/schannel.c [hunk 8]
```diff
@@ -672,12 +691,13 @@ schannel_connect_step2(struct connectdata *conn, int sockindex)
     infof(data, "schannel: SSL/TLS handshake complete\n");
   }
 
-#ifdef _WIN32_WCE
-  /* Windows CE doesn't do any server certificate validation.
-     We have to do it manually. */
-  if(conn->ssl_config.verifypeer)
+  if(conn->ssl_config.verifypeer &&
+     (DEFAULT_MANUAL_CRED_VALIDATION ||
+     (conn->ssl_config.CAfile && CAFILE_SUPPORTED))) {
```
Comment (mback2k) [inline lib/vtls/schannel.c]: Instead of checking these conditions here again, I would advise to use a variable to re-use the result of the checks in https://github.com/curl/curl/pull/1325/files#diff-2ba9d5898aad590d89cd83e17ad244fbL210. Or maybe just check if schannel_cred.dwFlags contains SCH_CRED_MANUAL_CRED_VALIDATION and conn->ssl_config.verifypeer is set.

File: lib/vtls/schannel.c [hunk 9]
```diff
@@ -1458,7 +1478,7 @@ int Curl_schannel_shutdown(struct connectdata *conn, int sockindex)
       failf(data, "schannel: ApplyControlToken failure: %s",
             Curl_sspi_strerror(conn, sspi_status));
 
-    host_name = Curl_convert_UTF8_to_tchar(hostname);
+    host_name = Curl_convert_UTF8_to_tchar(conn->host.name);
```
Comment (mback2k) [inline lib/vtls/schannel.c]: Again, please check that this change still supports HTTPS-proxies.

File: lib/vtls/schannel.c [hunk 10]
```diff
@@ -1574,7 +1594,332 @@ CURLcode Curl_schannel_random(unsigned char *entropy, size_t length)
   return CURLE_OK;
 }
 
-#ifdef _WIN32_WCE
+#ifdef HAVE_EXCLUSIVE_TRUST_MODE
+static CURLcode add_certs_to_store(HCERTSTORE trust_store,
```
Comment (mback2k) [inline lib/vtls/schannel.c]: Since schannel.c is slowly growing in number of lines, I would suggest moving all the custom certificate handling into a separate helper source file, e.g. schannel_cert.c. What do you think, @bagder and @jay?

File: lib/vtls/schannel.c [hunk 11]
```diff
@@ -1574,7 +1594,332 @@ CURLcode Curl_schannel_random(unsigned char *entropy, size_t length)
...
+              ca_file,
+              Curl_strerror(conn, GetLastError()));
+        result = CURLE_SSL_CACERT_BADFILE;
+        break;
+      }
+      if(bytes_read == 0) {
+        /* Premature EOF -- adjust the bufsize to the new value */
+        ca_file_bufsize = total_bytes_read;
+      }
+      else{
```
Comment (mback2k) [inline lib/vtls/schannel.c]: Please add a space between the else and { to stay in line with the rest of schannel.c

File: lib/vtls/schannel.c [hunk 12]
```diff
@@ -1574,7 +1594,332 @@ CURLcode Curl_schannel_random(unsigned char *entropy, size_t length)
...
+    END_CERT_LEN = strlen(END_CERT);
+
+    more_certs = 1;
+    current_ca_file_ptr = ca_file_buffer;
+    while(more_certs && current_ca_file_ptr != '\0') {
+      char *begin_cert_ptr = strstr(current_ca_file_ptr, BEGIN_CERT);
+      if(!begin_cert_ptr) {
+        more_certs = 0;
+      }
+      else{
```
Comment (mback2k) [inline lib/vtls/schannel.c]: See above, add space.

File: lib/vtls/schannel.c [hunk 13]
```diff
@@ -1574,7 +1594,332 @@ CURLcode Curl_schannel_random(unsigned char *entropy, size_t length)
...
+        char *end_cert_ptr = strstr(begin_cert_ptr, END_CERT);
+        if(!end_cert_ptr) {
+          failf(data,
+                "schannel: CA file is not correctly formatted '%s'",
+                ca_file);
+          result = CURLE_SSL_CACERT_BADFILE;
+          more_certs = 0;
+        }
+        else{
```
Comment (mback2k) [inline lib/vtls/schannel.c]: See above, add space.

File: lib/vtls/schannel.c [hunk 14]
```diff
@@ -1574,7 +1594,332 @@ CURLcode Curl_schannel_random(unsigned char *entropy, size_t length)
...
+          CERT_CONTEXT *cert_context = NULL;
+          BOOL add_cert_result = FALSE;
+          DWORD actual_content_type = 0;
+          DWORD cert_size = (DWORD)
+            ((end_cert_ptr + END_CERT_LEN) - begin_cert_ptr);
+
+          cert_blob.pbData = (BYTE *)begin_cert_ptr;
+          cert_blob.cbData = cert_size;
+          if(!CryptQueryObject(CERT_QUERY_OBJECT_BLOB,
+                                &cert_blob,
```
Comment (mback2k) [inline lib/vtls/schannel.c]: Please fix the indentation here.

File: lib/vtls/schannel.c [hunk 15]
```diff
@@ -1574,7 +1594,332 @@ CURLcode Curl_schannel_random(unsigned char *entropy, size_t length)
...
+                                NULL,
+                                &cert_context)) {
+
+            failf(data,
+                  "schannel: failed to extract certificate from CA file: %s",
+                  Curl_strerror(conn, GetLastError()));
+            result = CURLE_SSL_CACERT_BADFILE;
+            more_certs = 0;
+          }
+          else{
```
Comment (mback2k) [inline lib/vtls/schannel.c]: See above, add space.

File: lib/vtls/schannel.c [hunk 16]
```diff
@@ -1574,7 +1594,332 @@ CURLcode Curl_schannel_random(unsigned char *entropy, size_t length)
...
+
+            /* Sanity check that the cert_context object is the right type */
+            if(CERT_QUERY_CONTENT_CERT != actual_content_type) {
+              failf(data,
+                    "schannel: unexpected content type when extracting "
+                    "certificate: %d", actual_content_type);
+              result = CURLE_SSL_CACERT_BADFILE;
+              more_certs = 0;
+            }
+            else{
```
Comment (mback2k) [inline lib/vtls/schannel.c]: See above, add space.

File: lib/vtls/schannel.c [hunk 17]
```diff
@@ -1574,7 +1594,332 @@ CURLcode Curl_schannel_random(unsigned char *entropy, size_t length)
...
+              CertFreeCertificateContext(cert_context);
+              if(!add_cert_result) {
+                failf(data,
+                      "schannel: failed to add certificate from CA file "
+                      "to certificate store: %s",
+                      Curl_strerror(conn, GetLastError()));
+                result = CURLE_SSL_CACERT_BADFILE;
+                more_certs = 0;
+              }
+              else{
```
Comment (mback2k) [inline lib/vtls/schannel.c]: See above, add space.

File: lib/vtls/schannel.c [hunk 18]
```diff
@@ -1574,7 +1594,332 @@ CURLcode Curl_schannel_random(unsigned char *entropy, size_t length)
...
+          }
+        }
+      }
+    }
+
+    if(result == CURLE_OK && !found_certs) {
+      infof(data,
+            "schannel: did not add any certificates from CA file\n");
+    }
+  }while(0);
```
Comment (mback2k) [inline lib/vtls/schannel.c]: See above, add space. And why is this whole block inside a do { ... } while(0); loop? Since I do not see any continue, I don't see the logic behind that. Please elaborate.

File: lib/vtls/schannel.c [hunk 19]
```diff
@@ -1574,7 +1594,332 @@ CURLcode Curl_schannel_random(unsigned char *entropy, size_t length)
...
+    if(result == CURLE_PEER_FAILED_VERIFICATION) {
+      failf(data,
+            "schannel: CertGetNameString() failed to match "
+            "connection hostname (%s) against server certificate names",
+            conn->host.name);
+    }
+    else if(result != CURLE_OK) {
+      failf(data, "schannel: server certificate name verification failed");
+    }
+  }while(0);
```
Comment (mback2k) [inline lib/vtls/schannel.c]: Same as before. I think you are doing this to break out and go to the cleanup part of the function. But in this case I would actually recommend to use goto, since this is used in libcurl for such purposes and actually describes what you are doing.

File: lib/vtls/schannel.c [hunk 20]
```diff
@@ -1574,7 +1594,332 @@ CURLcode Curl_schannel_random(unsigned char *entropy, size_t length)
...
+    /* Determine the size of the string needed for the cert hostname */
+    len = CertGetNameString(pCertContextServer,
+                            CERT_NAME_DNS_TYPE,
+                            name_flags,
+                            NULL,
+                            NULL,
+                            0);
+    if(len == 0) {
+      failf(data,
+          "schannel: CertGetNameString() returned no "
```
Comment (mback2k) [inline lib/vtls/schannel.c]: Please fix the indentation here, thanks.

File: lib/vtls/schannel.c [hunk 21]
```diff
@@ -1574,7 +1594,332 @@ CURLcode Curl_schannel_random(unsigned char *entropy, size_t length)
...
+      if(!cert_hostname) {
+        result = CURLE_OUT_OF_MEMORY;
+      }
+      else{
+        int match_result;
+
+        match_result = Curl_cert_hostcheck(cert_hostname, conn->host.name);
+        if(match_result == CURL_HOST_MATCH) {
+          infof(data,
+              "schannel: connection hostname (%s) validated "
```
Comment (mback2k) [inline lib/vtls/schannel.c]: Please fix the indentation here, thanks.

File: lib/vtls/schannel.c [hunk 22]
```diff
@@ -1574,7 +1594,332 @@ CURLcode Curl_schannel_random(unsigned char *entropy, size_t length)
...
+              "against certificate name (%s)\n",
+              conn->host.name,
+              cert_hostname);
+          result = CURLE_OK;
+        }
+        else{
```
Comment (mback2k) [inline lib/vtls/schannel.c]: As before, add a space please.

File: lib/vtls/schannel.c [hunk 23]
```diff
@@ -1591,18 +1941,68 @@ static CURLcode verify_certificate(struct connectdata *conn, int sockindex)
                                             SECPKG_ATTR_REMOTE_CERT_CONTEXT,
                                             &pCertContextServer);
 
-  if((status != SEC_E_OK) || (pCertContextServer == NULL)) {
-    failf(data, "schannel: Failed to read remote certificate context: %s",
+  if((status != SEC_E_OK) || !pCertContextServer) {
+    failf(data, "schannel: failed to read remote certificate context: %s",
```
Comment (mback2k) [inline lib/vtls/schannel.c]: Please keep changes irrelevant for this new feature in a separate commit. As far as I can tell this is just a code-style change, right?

**Comment by mcnulty** [comment]
Thanks for your comments @mback2k! I hope to be able to address them within a week or so.

File: lib/vtls/schannel.c (hunk 2 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: Per reply above, this variable has been removed.

File: lib/vtls/schannel.c (hunk 3 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: Per reply above, this variable has been removed.

File: lib/vtls/schannel.c (hunk 4 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: Per reply above, this variable has been removed.

File: lib/vtls/schannel.c (hunk 5 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: Fixed.

File: lib/vtls/schannel.c (hunk 6 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: Good catch, this change has been reverted.

My changes were originally developed before HTTPS proxy support landed in master. When I was rebasing these changes onto master, I ran into some conflicts. In the process of resolving those changes, I think I inadvertently made this change.

File: lib/vtls/schannel.c (hunk 9 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: Per my comment above, this change has been reverted.

File: lib/vtls/schannel.c (hunk 10 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: I'd be glad to attempt that change.

Any tips on where this new file would need to be added in the build system?

I've been only using the winbuild makefiles. AFAICT, those makefiles won't need to be updated to account for the new source file.

A quick grep shows `schannel.c` referenced in:

```
Makefile
Makefile.in
lib/Makefile
lib/Makefile.in
lib/Makefile.inc
packages/Symbian/group/libcurl.mmp
```

Will the Visual Studio project files need to be updated?

File: lib/vtls/schannel.c (hunk 11 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: Fixed.

File: lib/vtls/schannel.c (hunk 12 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: Fixed.

File: lib/vtls/schannel.c (hunk 13 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: Fixed.

File: lib/vtls/schannel.c (hunk 14 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: Fixed.

File: lib/vtls/schannel.c (hunk 15 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: Fixed.

File: lib/vtls/schannel.c (hunk 16 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: Fixed.

File: lib/vtls/schannel.c (hunk 17 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: Fixed.

File: lib/vtls/schannel.c (hunk 18 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: This has been changed to a goto + cleanup label instead of the do while loop.

File: lib/vtls/schannel.c (hunk 21 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: Fixed.

File: lib/vtls/schannel.c (hunk 20 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: Fixed.

File: lib/vtls/schannel.c (hunk 22 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: Fixed.

File: lib/vtls/schannel.c (hunk 19 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: This has been changed to use a goto instead.

File: lib/vtls/schannel.c (hunk 23 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: Good catch, this change has been reverted.

File: lib/vtls/schannel.c (hunk 7 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: Per my comment above, this change has been reverted.

File: lib/vtls/schannel.c (hunk 8 above)
Comment (mcnulty) [inline lib/vtls/schannel.c]: Thanks for this suggestion, I went ahead added the flag, `use_manual_cred_validation`,  to ssh_connect_data for Schannel to store the result of checking these conditions in schannel_connect_step1.

This check is now:
`if(conn->ssl_config.verifypeer && connssl->use_manual_cred_validation) {`

I figured this would be acceptable since a similar approach is used for the `use_alpn` flag.

**Comment by mcnulty** [comment]
Just checking in on this PR. I think I addressed all of @mback2k's comments. Is there anything more I can do to help shepherd this change along?

**Comment by bagder** [comment]
Can you please squash the commits (and rebase them) so that they are the minimum set of logical changes required to provide this feature? It makes them easier for a final review and the rebase/ṕush will make them run through the (now) updated CI tests. (and I too would like @mback2k's ack on this)

**Comment by dscho** [comment]
@mcnulty the conflicts are most likely due to my recent changes to support switching the SSL backend at runtime. Would you like me to help you resolving them?

**Comment by mcnulty** [comment]
This PR has been rebased on a master.

@dscho: Thanks for the offer to help resolve the conflicts. I think I was successfully able to resolve the conflicts. It appears all that needed to be done was to define backend-specific data in the `ssl_backend_data` struct instead of in the `ssl_connect_data` struct and then utilize the BACKEND macro in schannel.c to access the backend-specific data. Does that align with your understanding?

**Comment by dscho** [comment]
> It appears all that needed to be done was to define backend-specific data in the ssl_backend_data struct instead of in the ssl_connect_data struct and then utilize the BACKEND macro in schannel.c to access the backend-specific data. Does that align with your understanding?

Yes, it does! Thanks!

**Comment by mcnulty** [comment]
Looks like the Travis [T=normal C=--enable-ares build](https://travis-ci.org/curl/curl/jobs/288080929) failed on test 1208:

> test 1208...[FTP PORT download, no data conn and no transient negative reply]
 1208: protocol FAILED:

File: lib/vtls/schannel.c [hunk 24]
```diff
@@ -291,21 +308,20 @@ schannel_connect_step1(struct connectdata *conn, int sockindex)
...
-        SCH_CRED_IGNORE_REVOCATION_OFFLINE;
-#else
-      schannel_cred.dwFlags = SCH_CRED_AUTO_CRED_VALIDATION;
-      /* TODO s/data->set.ssl.no_revoke/SSL_SET_OPTION(no_revoke)/g */
-      if(data->set.ssl.no_revoke)
-        schannel_cred.dwFlags |= SCH_CRED_IGNORE_NO_REVOCATION_CHECK |
+      if(BACKEND->use_manual_cred_validation) {
+        schannel_cred.dwFlags = SCH_CRED_MANUAL_CRED_VALIDATION |
+                                SCH_CRED_IGNORE_NO_REVOCATION_CHECK |
+                                SCH_CRED_IGNORE_REVOCATION_OFFLINE;
```
Comment (jay) [inline lib/vtls/schannel.c]: this I'm not fond of, windows users have the expectation that revocation checks are enabled by default. wouldn't we want to do the same for the certificate bundle they supply for consistency? (I'm assuming here that is what will happen since manual cred is now used for the bundle)

File: lib/vtls/schannel.c (hunk 10 above)
Comment (jay) [reply lib/vtls/schannel.c]: I will echo this , I think the verify stuff could be broken away schannel_verify/cert/whatever.c like how polarssl threadlock stuff went in polarssl_threadlock.c instead

File: tests/data/Makefile.inc [hunk 25]
```diff
@@ -55,7 +55,7 @@ test280 test281 test282 test283 test284 test285 test286 test287 test288 \
 test289 test290 test291 test292 test293 test294 test295 test296 test297 \
 test298 test299 test300 test301 test302 test303 test304 test305 test306 \
 test307 test308 test309 test310 test311 test312 test313 test314 test315 \
-                                test320 test321 test322 test323 test324 \
+test316 test317                 test320 test321 test322 test323 test324 \
```
Comment (jay) [inline tests/data/Makefile.inc]: i rebased this and renamed the tests to avoid conflicts 314 => 316 and 315 => 317

File: lib/config-win32.h [hunk 26]
```diff
@@ -612,6 +612,14 @@ Vista
 #  endif
 #endif
 
+/* Define whether the CERT_CHAIN_ENGINE_CONFIG structure has support for
+ * setting exclusive trust stores. This is used to implement the CAINFO easy
+ * option for Schannel.
+ */
+#if defined(_WIN32_WINNT) && (_WIN32_WINNT >= 0x0601)
+#define HAVE_EXCLUSIVE_TRUST_MODE 1
```
[the remark this answers is not in the document]
Comment (mcnulty) [reply lib/config-win32.h]: Interesting. Perhaps the CERT_CHAIN_ENGINE_CONFIG definition has changed in more recent versions of the Windows SDK?

Here's what the definition looks like in the 8.0 SDK:
```
typedef struct _CERT_CHAIN_ENGINE_CONFIG {

    DWORD       cbSize;
    HCERTSTORE  hRestrictedRoot;
    HCERTSTORE  hRestrictedTrust;
    HCERTSTORE  hRestrictedOther;
    DWORD       cAdditionalStore;
    HCERTSTORE* rghAdditionalStore;
    DWORD       dwFlags;
    DWORD       dwUrlRetrievalTimeout;      // milliseconds
    DWORD       MaximumCachedCertificates;
    DWORD       CycleDetectionModulus;

#if (NTDDI_VERSION >= NTDDI_WIN7)
    HCERTSTORE  hExclusiveRoot;
    HCERTSTORE  hExclusiveTrustedPeople;
#endif

#if (NTDDI_VERSION >= NTDDI_WIN8)
    DWORD       dwExclusiveFlags;
#endif

} CERT_CHAIN_ENGINE_CONFIG, *PCERT_CHAIN_ENGINE_CONFIG;
```

When I explicitly set _WIN32_WINNT to 0x0600 (Windows Server 2008), the hExclusiveRoot field is not present in the structure. Given this, I don't think this check can be done at runtime.

But, I do notice that other ifdef checks specific to Schannel are in the schannel.c file, for example, HAS_ALPN. Given that, I can move this check into schannel.c (or schannel_verify.c) instead of having it live in config-win32.h.

File: lib/vtls/schannel.c [hunk 27]
```diff
@@ -158,16 +158,19 @@ struct ssl_backend_data {
...
+  bool use_manual_cred_validation; /* true if manual cred validation is used */
 };
 
 #define BACKEND connssl->backend
 
 static Curl_recv schannel_recv;
 static Curl_send schannel_send;
 
-#ifdef _WIN32_WCE
+#define MAX_CAFILE_SIZE 67108864 /* 64 MB */
```
[the remark this answers is not in the document]
Comment (mcnulty) [reply lib/vtls/schannel.c]: Good point. Would a smaller value like 512k or 1M work? I'll go with 1M unless there are objections to that.

File: lib/vtls/schannel.c (hunk 24 above)
Comment (mcnulty) [reply lib/vtls/schannel.c]: Great catch! Looks like revocation checks were being unconditionally disabled for CE, regardless of the setting of the no_revoke flag.

Interestingly enough, it looks like the call to CertGetCertificateChain overrides these flags by setting the CERT_CHAIN_REVOCATION_CHECK_CHAIN flag in the flags for that function if the no_revoke flag is set.

That said, I'll still update this code to set dwFlags based on the no_revoke flag.

File: lib/vtls/schannel.c (hunk 10 above)
Comment (mcnulty) [reply lib/vtls/schannel.c]: Sounds good, I'll attempt that change.

File: tests/data/Makefile.inc (hunk 25 above)
Comment (mcnulty) [reply tests/data/Makefile.inc]: Cool, thanks!

**Comment by mcnulty** [comment]
Thanks for your review @jay! I think I've addressed all of your comments with my most recent push. 

> I think we should consider making it so that the curl tool only passes the CAINFO option in the case of schannel when the user has explicitly requested it, and not use autodetect, that way we could still preserve backwards compatibility. Thoughts?

I added some code to tool_operate.c in the operate_do function to skip the autodetect search if the SSL backend is Schannel. I used the CURLINFO_TLS_SSL_PTR option for curl_easy_getinfo to determine the SSL backend in use. I'm not sure if this was the best way to determine the SSL backend, but it seemed to do the trick.

**Comment by mcnulty** [comment]
I went ahead and rebased this PR on master. Any chance these changes could be considered for inclusion in 7.60.0?

**Comment by jay** [comment]
Thanks Dan I will review again when I can.

**Comment by jay** [comment]
Thanks. I changed the OS check from a build time check to a runtime check since it's common to build libcurl targeting an OS earlier than Windows 7, and yet we would still like this ability on Windows 7+ for those builds.
